doc/man3/EZ_Enable.3
'\"
'\" Copyright (c) 1997 Maorong Zou
'\" 
.TH EZ_Enable 3 "" EZWGL "EZWGL Functions"
.BS
.SH NAME
EZ_Enable, EZ_Disable \- turn on/off graphics capabilities

.SH SYNOPSIS
.nf
.B #include <EZ.h>
.sp
.BI "void EZ_Enable(int " mode )
.BI "void EZ_Disable(int " mode )

.SH ARGUMENTS
\fImode\fR Specifies a symbolic constant that represents a graphic
library capability. 

.SH DESCRIPTION
\fBEZ_Enable/EZ_Disable\fR turns on/off certain graphics library
capabilities. These capabilities are represented by the following
symbolic constants:
.sp
.in +5
\fBEZ_LIGHT[0-7]\fR  enable/disable light source 0--7.
.sp
\fBEZ_LIGHTING\fR enable/disable lighting calculations.
.sp
\fBEZ_NORMALIZE\fR enable/disable automatic normal vector normalization.
.sp
\fBEZ_COLOR_MATERIAL\fR enable/disable color material.
.sp
\fBEZ_CULL_FACE\fR enable/disable back face culling.
.sp
\fBEZ_FOG\fR enable/disable fog.
.sp
\fBEZ_DEPTH_TEST\fR enable/disable zbuffer.
.sp
\fBEZ_DITHER\fR enable/disable color dithering.
.in -5

.SH "SEE ALSO"
EZ_Color3f(3), EZ_Normal3f(3), EZ_Vertex3f(3), EZ_ColorMaterial(3)





ChangeLog
Fri Dec 10 23:44:16 1999  m z  <zou@hitachi1.ma.utexas.edu>
        
        * Added EZ_SSheetSortRegionR, EZ_SSheetSortColumns/Rows
        
1999-12-04  Maorong Zou  <mzou@linux115.ma.utexas.edu>

        * Released version 1.50 
        * added a CVS repository entry 
        *  :pserver:anonymous@cvs.math.utexas.edu:/home/cvs/EZWGL
        * and a read-only cvs web interface
        *  http://cvs.math.utexas.edu/cgi-bin/cvswe.cgi?cvsroot=EZWGL

1999-11-18  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* Added THUMB_WHEEN and SPREAD_SHEET widget
	* a few modifications in in FigItem ...

1999-09-22  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* EZ_RESOURCE_STRING resource
	* EZ_LedLoadPixelMatrix() EZ_LedLoadOnOffMatrix()
	* EZ_AllowDualLabel()
	* EZ_FlashWidgetBorder2()
	* NewFile  EZ_ResourceEdit.c
	* Incompatible change:  EZ_LED_PIXEL_SIZE now
	*        takes a pair of ints (xsize, ysize)

	
Sun Aug 22 01:26:59 1999  m z  <mzou@crab.ma.utexas.edu>

        * Separator Widget

Tue Aug 17 02:10:03 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added   EZ_SlideOpenWidget,  EZ_SlideCloseWidget
        *         EZ_SlideOpenWidgetRelativeTo
        * new widget EZ_WIDGET_HISTOGRAM

Tue Aug 10 12:33:26 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added the pare EZ_FreezeDisplay() and EZ_ThawDisplay()
        * New File EZ_ResourceEdit.c (resource editor functions)
        * Added EZ_RegisterApplicationResources(n,spec,config)
        * Modified all resource handling functions
        * changed EZ_Initialize and EZ_InitializeXrm from void to int
        *   (return -1 on failure, 0 on success)
        
Mon Jul 12 02:13:37 1999  m z  <mzou@crab.ma.utexas.edu>

        * New widget GradientBar
        * Fix a bug in EZ_PixelValue2RGB
        
Tue Jul  6 02:17:29 1999  m z  <mzou@crab.ma.utexas.edu>

        * Fix buffer overflow bug in EZ_Xpm.c
        * added line/word wrap for textWidget, IN BROWSING FILE MODE ONLY!!
        * (need to rewrite textWidget, the current line based
        *  implementation is a design flaw)

Thu Jun 17 01:49:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * New resources: SHAPE_PIXMAP, SHAPE_FILE
        * EZ_Bitmap *EZ_CreateShapeMaskFromFigure(fitem) EZ_Item *fitem;
        * Pixmap EZ_MakePixmapFromFigItem(item, width, height)
             EZ_Item *item; int *width, *height;
        
Fri May 14 23:30:19 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added a new tree type: branched-list-tree

Thu May 13 01:50:55 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Widget:  ScrollButton, SpinButton, Locator

Sun May  9 02:03:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added UDF widget resource handling.
        * EZ_RegisterResourcesAll(type, nresrc, resrc, configure);
        * EZ_RegisterResourcesWidget(widget, nresrc, resrc, configure);        

Tue May  4 23:36:31 1999  m z  <mzou@crab.ma.utexas.edu>

        *  Widget (horizontal/vertical) ruler
        *  
        *  Configuration options:
        *  EZ_HIGHLIGHT_BACKGROUND, EZ_HIGHLIGHT_FOREGROUND,
        *  EZ_OVERRIDE_REDIRECT

Tue May  4 01:22:57 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Function 
        * void  EZ_ConstrainFreeLabel(EZ_Widget *widget, int flag);
        * void  EZ_MoveWidgetWindow(EZ_Widget *widget, int x, int y);
        * void  EZ_ResizeWidgetWindow(EZ_Widget *widget, int w, int h);
        * void  EZ_MoveResizeWidgetWindow(EZ_Widget *widget, int x, int y, int w, int h);        
        * Added optimization for manipulating images.
        *
        
Sat Apr 17 01:25:25 1999  m z  <mzou@crab.ma.utexas.edu>

        * Autoconf (finally)

1999-04-16  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * added a vt100 emulator widget (EZ_WIDGET_TERM), derived from xiterm.
        * added support for image format: png and tiff
        
Tue Dec  8 11:39:04 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        *  new function:     EZ_ScaleRGBT
        *                    EZ_XPixmap2RGB
        *                    EZ_SelectInput
        *                    EZ_CreateBackgroundPixmapsFromRawRGB

Sat Dec  5 03:06:58 1998  m z  <mzou@bob.ma.utexas.edu>

        * EZ_LABEL_SHADOW  "labelShadow" config. option

Thu Dec  3 14:54:21 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * new function:   EZ_SetDirTreeSorter
        *                 EZ_MarkWidgetDirty
        
Mon Nov 30 12:25:50 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * EZ_WM_IGNORE  "wmIgnore"    configuration option
        *
        * new function    EZ_GetIconTextGeometry
        *                 EZ_SetGlobalEventFilter(filter, data)
        
Thu Nov 19 16:35:47 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	*  EZ_TILE_ORIGIN  tileOrigin: mode x y
	*  resources.  mode=0,1,2    (0 for off)

	*  It is now possible to make all EZ widgets
	*  pseudo transparent under Enlightment WM ...

	*  A few image readers have been modified to
	*  return RGB images.

	*  new functions EZ_ScaleRGB, EZ_ReadImageFile,
	*                EZ_ScaleLabelPixmap,
	*                EZ_SwapLabelPixmaps, 
	*                EZ_CopyLabelPixmaps,
	*                EZ_TraverseAllWidgets,

	*  In EZ_DisplayWidget, the order for displaying widget tree
	*  is reversed. Now widgets refresh a lot faster visually.
	
	*  Bug Fix:  whe a radio button is destroyed, the widget is
	*            not removed from its button group immediately.



lib/EZ_Enable.c
/*********************************************************************
 *
 *         EZWGL, the EZ Widget and Graphics Library
 *
 * *             Copyright (C) 1996, 1997, 1998  Maorong Zou
 *  
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 **********************************************************************/
/*
 *  June 1996.  Beta Release.
 *  Sept 1996.  Release Version 1.0
 *  Dec 1996.  Release Version 1.1 Beta
 *  April 1997.  Release Version 1.2
 *  November 1997.  Release Version 1.3
 *  November 1998.  Release Version 1.4
 *  December 1999.  Release Version 1.50
 */
/**************************************************************
 *
 *    Enable/Disable
 *
 */
#define _EZ_ENABLE_C_
#include "EZ_Widget.h"

/**************************************************************
 *
 *   Functions implemented in this file
 */
void         EZ_Enable MY_ANSIARGS((int mode));
void         EZ_Disable MY_ANSIARGS((int mode));

/**************************************************************
 *
 * local functions
 */

/**************************************************************/
void EZ_Enable(mode)
     int mode;
{
  if(EZ_InCompilingMode != 0)
    {
      EZ_Token *t = AllocateNTokens(2);
      t->i = TOKEN_ENABLE;
      t++;
      t->i = mode;
    }
  if(EZ_InExecutionMode != 0)
    {  
      switch(mode)
	{
	case EZ_LIGHT0:
	case EZ_LIGHT1:
	case EZ_LIGHT2:
	case EZ_LIGHT3:
	case EZ_LIGHT4:
	case EZ_LIGHT5:
	case EZ_LIGHT6:
	case EZ_LIGHT7:
	  {
	    int last = -1, j, i = mode - EZ_LIGHT0;
	    
	    EZ_LightSrcsMask[i] = 1;
	    for(i = 0; i < 8; i++)
	      if(EZ_LightSrcsMask[i]) last = i;
	    EZ_LastEnabledLight  = last;	    
	    
	    if(EZ_LightingModel.localviewer == 0)
	      {
		j = 0;
		for(i = 0; i < 8; i++)
		  {
		    if(EZ_LightSrcsMask[i])
		      {
			if((EZ_CurrentLightSrc[i].position)[3] != 0)
			  {
			    EZ_LightIsInfinite = 0;
			    j = 1;
			    break;
			  }
		      }
		  }
		if(j == 0) EZ_LightIsInfinite = 1;
	      }
	  }
	  /* may turn on new light sources */
	  EZ_LColorValid = 0;	  
	  EZ_LColorValidB = 0;	  
	  break;
	case EZ_LIGHTING:
	  EZ_LightingEnabled = 1;
	  EZ_LColorValid = 0;	  
	  EZ_LColorValidB = 0;	  
	  break;
	case EZ_NORMALIZE:
	  EZ_NormalizeEnabled = 1;
	  break;
	case EZ_COLOR_MATERIAL:
	  EZ_ColorMaterialEnabled = 1;
	  EZ_FColorValid = 0;	  
	  EZ_BColorValid = 0;
	  break;
	case EZ_CULL_FACE:
	  EZ_Cullface = 1;
	  break;
	case EZ_DITHER:
	  if(EZ_Depth <= 8)  /* dithering is only available in 8bit RGB mode */
	    EZ_DitheringFlag = 1; 
	  break;
	case EZ_FOG:
	  EZ_FogEnabled = 1;
	  break;
	case EZ_DEPTH_TEST:
	  EZ_DepthTestEnabled = 1;
	  EZ_ZBufferSet = 1;
	  EZ_GLConfig();
	  break;
	default:
	  EZ_GLError("EZ_Enable", "Unknown state");
	  break;
	}
    }
}

void EZ_Disable(mode)
     int mode;
{
  if(EZ_InCompilingMode != 0)
    {
      EZ_Token *t = AllocateNTokens(2);
      t->i = TOKEN_DISABLE;
      t++;
      t->i = mode;
    }
  if(EZ_InExecutionMode != 0)
    {  
      switch(mode)
	{
	case EZ_LIGHT0:
	case EZ_LIGHT1:
	case EZ_LIGHT2:
	case EZ_LIGHT3:
	case EZ_LIGHT4:
	case EZ_LIGHT5:
	case EZ_LIGHT6:
	case EZ_LIGHT7:
	  {
	    int last = -1, j, i = mode - EZ_LIGHT0;
	    EZ_LightSrcsMask[i] = 0;
	    
	    for(i = 0; i < 8; i++)
	      if(EZ_LightSrcsMask[i]) last = i;
	    EZ_LastEnabledLight  = last;

	    if(EZ_LightingModel.localviewer == 0)
	      {
		j = 0;
		for(i = 0; i < 8; i++)
		  {
		    if(EZ_LightSrcsMask[i])
		      {
			if((EZ_CurrentLightSrc[i].position)[3] != 0)
			  {
			    EZ_LightIsInfinite = 0;
			    j = 1;
			    break;
			  }
			last = i;
		      }
		  }
		if(j == 0) EZ_LightIsInfinite = 1;
	      }
	  }
	  /* turn off lights affects the 'current color' */
	  EZ_FColorValid = 0;	  
	  EZ_BColorValid = 0;	  
	  break;
	case EZ_LIGHTING:
	  EZ_LightingEnabled = 0;
	  EZ_FColorValid = 0;	  
	  EZ_BColorValid = 0;	  
	  break;
	case EZ_NORMALIZE:
	  EZ_NormalizeEnabled = 0;
	  break;
	case EZ_COLOR_MATERIAL:
	  EZ_ColorMaterialEnabled = 0;
	  EZ_FColorValid = 0;	  
	  EZ_BColorValid = 0;	  
	  break;
	case EZ_CULL_FACE:
	  EZ_Cullface = 0;
	  break;
	case EZ_DITHER:
	  EZ_DitheringFlag = 0;
	  break;
	case EZ_FOG:
	  EZ_FogEnabled = 0;
	  break;
	case EZ_DEPTH_TEST:
	  EZ_DepthTestEnabled = 0;
	  EZ_SelectRenderFunctions(EZ_CurrentGWindow);
	  break;
	default:
	  EZ_GLError("EZ_Disable", "Unknown state");
	  break;
	}
    }
}
/**********************************************************************/
#undef _EZ_ENABLE_C_
